src/main/scala/zio/zmx/prometheus/PrometheusDataModel.scala

```diff
@@ -0,0 +1,136 @@
+      override def buckets: List[Double] = 0.until(count).map(i => start * Math.pow(factor, i.toDouble)).toList
+    }
+
+  }
+
+  final case class Quantile(
+    phi: Double,  // The quantile
+    error: Double // The error margin
+  )
+
+  final case class Counter(name: String, help: Option[String], labels: Map[String, String], count: Double)
+      extends Metric {
+    // Must haves
+    def inc: Counter                    = copy(count = count + 1)
+    def inc(d: Double): Option[Counter] = if (d >= 0) Some(copy(count = count + d)) else None
+
+    // May haves
+    def reset: Counter = copy(count = 0)
+
+    // Encouraged
+    // Dow we want to count exceptions / if so, how
+  }
+
+  // NOTE: keep smart constructor; instruct user to use it
+  def counter(name: String, help: Option[String], labels: Map[String, String]) =
+    Counter(name, help, labels, count = 0)
+
+  final case class Gauge(name: String, help: Option[String], labels: Map[String, String], value: Double)
+      extends Metric {
+    // Must haves
+    def inc: Gauge            = inc(1)
+    def inc(v: Double): Gauge = copy(value = value + v)
+    def dec: Gauge            = inc(-1)
+    def dec(v: Double): Gauge = inc(-v)
+    def set(v: Double): Gauge = copy(value = v)
+
+    // Should haves
+    def setToCurrentTime(): Gauge = copy(value = java.time.ZonedDateTime.now().toEpochSecond().toDouble)
+
+    // Encouraged
+    // start / stop timer
+  }
+
+  def gauge(name: String, help: Option[String], labels: Map[String, String])                  = Gauge(name, help, labels, value = 0)
+  def gauge(name: String, help: Option[String], labels: Map[String, String], startAt: Double) =
+    Gauge(name, help, labels, value = startAt)
+
+  /* Requirements:
+   * A histogram MUST NOT allow le as a user-set label, as le is used internally to designate buckets.
+   * A histogram MUST offer a way to manually choose the buckets. Ways to set buckets in a linear(start, width, count) and exponential(start, factor, count) fashion SHOULD be offered. Count MUST exclude the +Inf bucket.
+   * A histogram SHOULD have the same default buckets as other client libraries. Buckets MUST NOT be changeable once the metric is created.
+   * A histogram MUST have the following methods:
+   *   - observe(double v): Observe the given amount
+   * A histogram SHOULD have the following methods:
+   * Some way to time code for users in seconds. In Python this is the time() decorator/context manager. In Java this is startTimer/observeDuration. Units other than seconds MUST NOT be offered (if a user wants something else, they can do it by hand). This should follow the same pattern as Gauge/Summary.
+   * Histogram _count/_sum and the buckets MUST start at 0.
+   */
+  final case class Histogram(
+    name: String,
+    help: Option[String],
+    labels: Map[String, String],
+    buckets: BucketType.Buckets,
+    sum: Double
```

Comment:
this structure makes me thing of algebird: 
* https://twitter.github.io/algebird/datatypes/approx/q_tree.html
* https://twitter.github.io/algebird/datatypes/approx/exponential_histogram.html

Reply:
The exponential histogram seems to be useful for the Prometheus summaries, but need to double check. The Q-Tree might be used as a Histogram, but need to get my head around how that would be set up. I am inclined to get a simple version up and running and leave Algebird usage for later.

Reply:
No pressure, it is just for the "state of the art" info 😉.